### PR TITLE
Add column policy_usage to aws_iam_policy table

### PR DESCRIPTION
**Pull request by cjsrkd3321**
# Integration test logs
<details>
  <summary>Logs</summary>
  
```
> make
go build -o  ~/.steampipe/plugins/hub.steampipe.io/plugins/turbot/aws@latest/steampipe-plugin-aws.plugin  *.go
> steampipe -v
steampipe version 0.10.0
> steampipe plugin list
+--------------------------------------------------+---------+-------------+
| Name                                             | Version | Connections |
+--------------------------------------------------+---------+-------------+
| hub.steampipe.io/plugins/turbot/aws@latest       | local   | aws         |
| hub.steampipe.io/plugins/turbot/steampipe@latest | 0.1.3   | steampipe   |
+--------------------------------------------------+---------+-------------+
```
</details>

# Example query results
<details>
  <summary>Results</summary>
  
```
> select name, policy_usage from aws_iam_policy where is_aws_managed is false
+---------------------------------------------------------------------------------------------------------------------------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| name                                                                                                                            | policy_usage                                                                                                                                                                           |
+---------------------------------------------------------------------------------------------------------------------------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| AWSControlTowerCloudTrailRolePolicy                                                                                             | {"IsTruncated":false,"Marker":null,"PolicyGroups":null,"PolicyRoles":[{"RoleId":"SECRET","RoleName":"AWSControlTowerCloudTrailRole"}],"PolicyUsers":null}                              |
| AWSLambdaBasicExecutionRole-912ec387-a0d6-4caa-bfc3-163d3567be4b                                                                | {"IsTruncated":false,"Marker":null,"PolicyGroups":null,"PolicyRoles":[{"RoleId":"SECRET","RoleName":"testtest-role-83s8u3w0"}],"PolicyUsers":null}                                     |
| CloudTrailPolicyForCloudWatchLogs_82e4c54d-ee0a-40ff-b7db-1b06d895b80c                                                          | {"IsTruncated":false,"Marker":null,"PolicyGroups":null,"PolicyRoles":[{"RoleId":"SECRET","RoleName":"cloudtrailrole"}],"PolicyUsers":null}                                             |
| AWSLambdaBasicExecutionRole-8eb2ae33-24a0-47d4-bed4-61beb0e1529f                                                                | {"IsTruncated":false,"Marker":null,"PolicyGroups":null,"PolicyRoles":[{"RoleId":"SECRET","RoleName":"eventbridgetest-role-0hmq5sxq"}],"PolicyUsers":null}                              |
| AWSControlTowerAdminPolicy                                                                                                      | {"IsTruncated":false,"Marker":null,"PolicyGroups":null,"PolicyRoles":[{"RoleId":"SECRET","RoleName":"AWSControlTowerAdmin"}],"PolicyUsers":null}                                       |
| AWSLambdaBasicExecutionRole-6ba5a28f-1dbe-49bf-b9bf-c2ca442eb788                                                                | {"IsTruncated":false,"Marker":null,"PolicyGroups":null,"PolicyRoles":[{"RoleId":"SECRET","RoleName":"test-role-rvxs2cxf"}],"PolicyUsers":null}                                         |
| AWSLambdaBasicExecutionRole-75c44454-4fb6-456f-bbcc-ddf7bad05f3a                                                                | {"IsTruncated":false,"Marker":null,"PolicyGroups":null,"PolicyRoles":[{"RoleId":"SECRET","RoleName":"jira-test-role-flrbia7j"}],"PolicyUsers":null}                                    |
| AWSLambdaBasicExecutionRole-d670d758-62c0-4b90-b2a0-1207f053437a                                                                | {"IsTruncated":false,"Marker":null,"PolicyGroups":null,"PolicyRoles":[{"RoleId":"SECRET","RoleName":"ec2-sg-eventbridge-test-role-kpw56vsr"}],"PolicyUsers":null}                      |
| AWSControlTowerStackSetRolePolicy                                                                                               | {"IsTruncated":false,"Marker":null,"PolicyGroups":null,"PolicyRoles":[{"RoleId":"SECRET","RoleName":"AWSControlTowerStackSetRole"}],"PolicyUsers":null}                                |
| aws-security-hub-automated-response-and-remediation-PlaybookServiceCatalog-14R9J8OT-SHARRCatalogUserPolicy58114AE9-3HQV00N9JLQ9 | {"IsTruncated":false,"Marker":null,"PolicyGroups":[{"GroupId":"SECRET","GroupName":"SO0111-SHARR_Catalog_User_ap-northeast-2"}],"PolicyRoles":null,"PolicyUsers":null}                 |
| AWSLambdaBasicExecutionRole-5d6522bb-d636-4050-b33e-7bd2f2bbe55c                                                                | {"IsTruncated":false,"Marker":null,"PolicyGroups":null,"PolicyRoles":null,"PolicyUsers":null}                                                                                          |
| aws-security-hub-automated-response-and-remediation-PlaybookServiceCatalog-14R9J8O-SHARRCatalogAdminPolicyE502F9A1-3WXBFGS8FNBC | {"IsTruncated":false,"Marker":null,"PolicyGroups":[{"GroupId":"SECRET","GroupName":"SO0111-SHARR_Catalog_Admin_ap-northeast-2"}],"PolicyRoles":null,"PolicyUsers":null}                |
| AWSLambdaBasicExecutionRole-11e3627b-35ab-4db4-8bdd-936d0bf890c6                                                                | {"IsTruncated":false,"Marker":null,"PolicyGroups":null,"PolicyRoles":[{"RoleId":"SECRET","RoleName":"testtest-role-l7mgva91"}],"PolicyUsers":null}                                     |
| AWSLambdaBasicExecutionRole-7716607f-c0e1-4d00-9f0f-faaa814be624                                                                | {"IsTruncated":false,"Marker":null,"PolicyGroups":null,"PolicyRoles":[{"RoleId":"SECRET","RoleName":"test1-role-r4car3dq"}],"PolicyUsers":null}                                        |
| AWSLambdaBasicExecutionRole-1160f530-5214-4f93-a2da-a1bae6d04e75                                                                | {"IsTruncated":false,"Marker":null,"PolicyGroups":null,"PolicyRoles":[{"RoleId":"SECRET","RoleName":"cloudtrail-test-role-w25a9j0l"}],"PolicyUsers":null}                              |
+---------------------------------------------------------------------------------------------------------------------------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
```
</details>

### CODE REVIEW

**Comment by rajlearner17** [comment]
@cjsrkd3321 Thanks for using Steampipe!! very glad to see this PR. We will review and take appropriate steps.

File: aws/table_aws_iam_policy.go [hunk 1]
```diff
@@ -233,6 +240,28 @@ func getPolicyVersion(ctx context.Context, d *plugin.QueryData, h *plugin.Hydrat
...
+	svc, err := IAMService(ctx, d)
+	if err != nil {
+		return nil, err
+	}
+
+	params := &iam.ListEntitiesForPolicyInput{
+		PolicyArn: policy.Arn,
+	}
+
+	usage, err := svc.ListEntitiesForPolicy(params)
```
Comment (bigdatasourav) [inline aws/table_aws_iam_policy.go:257]: @cjsrkd3321 could you please use API pagination -  ListEntitiesForPolicyPages

File: aws/table_aws_iam_policy.go [hunk 2]
```diff
@@ -109,6 +109,13 @@ func tableAwsIamPolicy(_ context.Context) *plugin.Table {
 				Hydrate:     getPolicyVersion,
 				Transform:   transform.FromField("PolicyVersion.Document").Transform(unescape).Transform(policyToCanonical),
 			},
+			{
+				Name:        "policy_usage",
+				Description: "Lists all IAM users, groups, and roles that the specified managed policy is attached to.",
+				Type:        proto.ColumnType_JSON,
+				Hydrate:	 getPolicyUsage,
```
Comment (bigdatasourav) [inline aws/table_aws_iam_policy.go:116]: Please update the format. you can run `gofmt -w filename.go`

Comment (cjsrkd3321) [reply aws/table_aws_iam_policy.go]: Sorry about that. I applied gofmt command.

File: aws/table_aws_iam_policy.go (hunk 1 above)
Comment (cjsrkd3321) [reply aws/table_aws_iam_policy.go]: modification completed.

File: aws/table_aws_iam_policy.go [hunk 3]
```diff
@@ -233,6 +240,37 @@ func getPolicyVersion(ctx context.Context, d *plugin.QueryData, h *plugin.Hydrat
...
+
+	params := &iam.ListEntitiesForPolicyInput{
+		PolicyArn: policy.Arn,
+	}
+
+	policyUsageData := &iam.ListEntitiesForPolicyOutput{}
+
+	err = svc.ListEntitiesForPolicyPages(
+		params,
+		func(page *iam.ListEntitiesForPolicyOutput, lastPage bool) bool {
```
Comment (bigdatasourav) [inline aws/table_aws_iam_policy.go:261]: @cjsrkd3321 as I can see you have first updated it as below and I believe that was the correct approach as with the current update, we are not using pagination.
```
policyUsageData := []*iam.ListEntitiesForPolicyOutput{}

	err = svc.ListEntitiesForPolicyPages(
		params,
		func(page *iam.ListEntitiesForPolicyOutput, lastPage bool) bool {
			policyUsageData = append(policyUsageData, page)
			return !lastPage
		},
	)
	
```

File: aws/table_aws_iam_policy.go [hunk 4]
```diff
@@ -233,6 +240,37 @@ func getPolicyVersion(ctx context.Context, d *plugin.QueryData, h *plugin.Hydrat
...
 }
 
+func getPolicyUsage(ctx context.Context, d *plugin.QueryData, h *plugin.HydrateData) (interface{}, error) {
+	plugin.Logger(ctx).Trace("getPolicyUsage")
+	policy := h.Item.(*iam.Policy)
+
+	// Create Session
+	svc, err := IAMService(ctx, d)
+	if err != nil {
+		return nil, err
```
Comment (bigdatasourav) [inline aws/table_aws_iam_policy.go:250]:
```suggestion
plugin.Logger(ctx).Error("getPolicyUsage", "service_connection", err)
		return nil, err
```
```suggestion
		return nil, err
```

File: aws/table_aws_iam_policy.go [hunk 5]
```diff
@@ -233,6 +240,37 @@ func getPolicyVersion(ctx context.Context, d *plugin.QueryData, h *plugin.Hydrat
...
+	err = svc.ListEntitiesForPolicyPages(
+		params,
+		func(page *iam.ListEntitiesForPolicyOutput, lastPage bool) bool {
+			policyUsageData = page
+			return !lastPage
+		},
+	)
+
+	if err != nil {
+		return nil, err
```
Comment (bigdatasourav) [inline aws/table_aws_iam_policy.go:268]:
```suggestion
plugin.Logger(ctx).Error("getPolicyUsage", "ListEntitiesForPolicyPages", err)
		return nil, err
```
```suggestion
		return nil, err
```

File: aws/table_aws_iam_policy.go (hunk 3 above)
Comment (bigdatasourav) [reply aws/table_aws_iam_policy.go:262]: @cjsrkd3321 Thanks for the response 👍 , I guess, I was not very clear in the above note. This is what we want you to consider again 
Your first [fix](https://github.com/turbot/steampipe-plugin-aws/pull/815/commits/a2372936fe1a4fdfbf71f53fb0594a4d7ecd3c17) after my suggestion was good

However, you changed the same in the [latest](https://github.com/turbot/steampipe-plugin-aws/pull/815/commits/5fcbdeb403abf3c1408b204fe5b5f001d336533e)

Pls do as below to retain your first change

```
func getPolicyUsage(ctx context.Context, d *plugin.QueryData, h *plugin.HydrateData) (interface{}, error) {
	plugin.Logger(ctx).Trace("getPolicyUsage")
	policy := h.Item.(*iam.Policy)

	// Create Session
	svc, err := IAMService(ctx, d)
	if err != nil {
		return nil, err
	}

	params := &iam.ListEntitiesForPolicyInput{
		PolicyArn: policy.Arn,
	}
	policyUsageData := []*iam.ListEntitiesForPolicyOutput{}

	err = svc.ListEntitiesForPolicyPages(
		params,
		func(page *iam.ListEntitiesForPolicyOutput, lastPage bool) bool {
			policyUsageData = append(policyUsageData, page)
			return !lastPage
		},
	)

	if err != nil {
		return nil, err
	}

	return policyUsageData, nil
}
```
Pls revert back if you need further help

Comment (cjsrkd3321) [reply aws/table_aws_iam_policy.go:260]: I finished revert and re-commit. At first fix's output is [{"PolicyGroups": [{"GroupName": "BLAH"}]}] but, actually output is [this](https://docs.aws.amazon.com/cli/latest/reference/iam/list-entities-for-policy.html#examples). therefore, I just modified code. anyway i respect your opinion thank you!

**Review by bigdatasourav — APPROVED** [review]
LGTM

**Comment by e-gineer** [comment]
I love the idea of making this data available in Steampipe, but am wondering if it would be better as a separate table?

The `aws_iam_policy` table is already a heavy / slow table to query, and this per-row Hydrate function with it's own pagination will definitely create a lot of API calls. `select * from aws_iam_policy` will now be dependent on these sub-listing / paginated calls for performance.

What if we moved this to be a separate `aws_iam_policy_usage` table?

**Comment by cjsrkd3321** [comment]
> I love the idea of making this data available in Steampipe, but am wondering if it would be better as a separate table?
> 
> The `aws_iam_policy` table is already a heavy / slow table to query, and this per-row Hydrate function with it's own pagination will definitely create a lot of API calls. `select * from aws_iam_policy` will now be dependent on these sub-listing / paginated calls for performance.
> 
> What if we moved this to be a separate `aws_iam_policy_usage` table?

I agree your comment. But, It's hard to work right away. basically, It would be nice to make it possible to search based on policy_arn. If you have other optinions, please comment.

**Pull request merged**